kopf/_core/intents/piggybacking.py

```diff
@@ -102,6 +116,63 @@ def login_via_client(
     )
 
 
+# This is basically a copy of `login_via_client` changed to use the
+# kubernetes_asyncio client library.
+async def login_via_async_client(
+        *,
+        logger: typedefs.Logger,
+        **_: Any,
+) -> Optional[credentials.ConnectionInfo]:
+
+    # Keep imports in the function, as module imports are mocked in some tests.
+    try:
+        import kubernetes_asyncio.config
+    except ImportError:
+        return None
+
+    try:
+        kubernetes_asyncio.config.load_incluster_config()  # cluster env vars
+        logger.debug("Async client is configured in cluster with service account.")
+    except kubernetes_asyncio.config.ConfigException as e1:
+        try:
+            await kubernetes_asyncio.config.load_kube_config()  # developer's config files
+            logger.debug("Async client is configured via kubeconfig file.")
+        except kubernetes_asyncio.config.ConfigException as e2:
+            raise credentials.LoginError("Cannot authenticate the async client library "
+                                         "neither in-cluster, nor via kubeconfig.")
+
+    # We do not even try to understand how it works and why. Just load it, and extract the results.
+    # For kubernetes client >= 12.0.0 use the new 'get_default_copy' method
+    if callable(getattr(kubernetes_asyncio.client.Configuration, 'get_default_copy', None)):
+        config = kubernetes_asyncio.client.Configuration.get_default_copy()
+    else:
+        config = kubernetes_asyncio.client.Configuration()
+
+    # For auth-providers, this method is monkey-patched with the auth-provider's one.
+    # We need the actual auth-provider's token, so we call it instead of accessing api_key.
+    # Other keys (token, tokenFile) also end up being retrieved via this method.
+    header: Optional[str] = config.get_api_key_with_prefix('BearerToken')
+    parts: Sequence[str] = header.split(' ', 1) if header else []
+    scheme, token = ((None, None) if len(parts) == 0 else
+                     (None, parts[0]) if len(parts) == 1 else
+                     (parts[0], parts[1]))  # RFC-7235, Appendix C.
+
+    # Interpret the config object for our own minimalistic credentials.
+    # Note: kubernetes client has no concept of a "current" context's namespace.
+    return credentials.ConnectionInfo(
+        server=config.host,
+        ca_path=config.ssl_ca_cert,  # can be a temporary file
+        insecure=not config.verify_ssl,
+        username=config.username or None,  # an empty string when not defined
+        password=config.password or None,  # an empty string when not defined
+        scheme=scheme,
+        token=token,
+        certificate_path=config.cert_file,  # can be a temporary file
+        private_key_path=config.key_file,  # can be a temporary file
+        priority=PRIORITY_OF_ASYNC_CLIENT,
+    )
```

Comment:
The only thing I would change would be to pull this out into a seperate method that takes a number of arguments (header, host, ssl_ca_cert, verify_ssl, username, password, cert_file, key_file and priority). Since the two `Configuration` classes are almost-but-not-quite identical, I don't think there's a good way unless you want to make a `typing.Protocol` class to represent both.

Reply:
Thanks for the feedback. But I don't see the point here. The code duplication is minimal and adding a function would basically just wrap the creation of the ConnectionInfo object.